### PR TITLE
Combined lock

### PR DESCRIPTION
**Pull request by joshua-cogliati-inl**
--------
Pull Request Description
--------
##### What issue does this change request address?
Closes https://github.com/idaholab/raven/issues/1881

##### What are the significant changes in functionality due to this change request?
Adds a lock in DistributedMemoryRunner to prevent two threads stepping on each other.
Locks more in JobHandler cleanJobQueue because DistributedMemoryRunner isDone can't be called while kill is also being called.
Also makes a variable in DistributedMemoryRunner private because it really shouldn't be touched by other classes.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: ravenframework/Runners/DistributedMemoryRunner.py [hunk 1]
```diff
@@ -66,25 +72,26 @@ def isDone(self):
...
+          try:
+            ray.get(self.__func, timeout=waitTimeOut)
+            return True
+          except ray.exceptions.GetTimeoutError:
+            return False
+          except ray.exceptions.RayTaskError:
+            #The code gets this undocumented error, and
+            # I assume it means the task has unfixably died,
+            # and so is done
+            return True
```
Comment (PaulTalbot-INL) [inline ravenframework/Runners/DistributedMemoryRunner.py:89]: `True` means the job probably ended successfully? Is that the right return here? Do we believe the job probably finished, but the task didn't terminate as expected?

File: tests/framework/CodeInterfaceTests/RAVEN/optimizer.xml [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+      Inner raven runs 100 MC samples, does basic stats, and returns the mean, sigma to the outer raven.
+    </description>
+    <analytic>
+      Each sample in this test converges towards the analytic solutions discussed in the "changing lower, upper bounds" section of
+      the "attenuate" analytic model documentation.
+    </analytic>
+    <revisions>
+      <revision author="talbpaul" date="2018-05-02">added "fully-correlated" variable for innerUpperBound as a test of the mechanic</revision>
+      <revision author="talbpaul" date="2018-11-06">modified inner run to use grid for speed, consistent results</revision>
+    </revisions>
```
Comment (PaulTalbot-INL) [inline tests/framework/CodeInterfaceTests/RAVEN/optimizer.xml:19]: I don't think these revisions apply to this new test.

File: tests/framework/CodeInterfaceTests/RAVEN/optimizer.xml [hunk 3]
```diff
@@ -0,0 +1,145 @@
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>framework/CodeInterfaceTests/RAVEN.Optimizer</name>
+    <author>talbpaul</author>
+    <created>2018-01-18</created>
```
Comment (PaulTalbot-INL) [inline tests/framework/CodeInterfaceTests/RAVEN/optimizer.xml:5]: the author and creation date should probably be updated

File: tests/framework/CodeInterfaceTests/RAVEN/tests [hunk 4]
```diff
@@ -5,6 +5,15 @@
...
    max_time = 500
  [../]
+ [./Optimizer]
+   type = 'RavenFramework'
+   input = 'optimizer.xml'
+   [./data]
+     type = OrderedCSV
+     output = 'Optimizer/opt_export_0.csv'
+   [../]
+   max_time = 500
```
Comment (PaulTalbot-INL) [inline tests/framework/CodeInterfaceTests/RAVEN/tests:15]: Does it really take this long? It doesn't look like it needed many optimization steps to find the solution on the boundary.

Comment (joshua-cogliati-inl) [reply tests/framework/CodeInterfaceTests/RAVEN/tests:15]: On the regression machines, the slowest I saw was 107 seconds, so I will update that.

File: ravenframework/Runners/DistributedMemoryRunner.py (hunk 1 above)
Comment (joshua-cogliati-inl) [reply ravenframework/Runners/DistributedMemoryRunner.py:89]: Hm, well, isDone can either return true or false, and if it returns false, then JobHandler will think that it is still running, and so not remove it.  Setting returnCode = -1 might be the correct thing to do here.

File: tests/framework/CodeInterfaceTests/RAVEN/optimizer.xml (hunk 2 above)
Comment (joshua-cogliati-inl) [reply tests/framework/CodeInterfaceTests/RAVEN/optimizer.xml:19]: Agreed.

File: ravenframework/Runners/DistributedMemoryRunner.py (hunk 1 above)
Comment (PaulTalbot-INL) [reply ravenframework/Runners/DistributedMemoryRunner.py:89]: Oh, okay, I misread, this is just telling if the guy is done nor not. `True` seems like a fair response.

Comment (joshua-cogliati-inl) [reply ravenframework/Runners/DistributedMemoryRunner.py:89]: (actually, this might be the cause of #1901 )

Comment (joshua-cogliati-inl) [reply ravenframework/Runners/DistributedMemoryRunner.py:90]: FYI: For future reference, the reason that returnCode = -1 needs to be set here as opposed to _collectRunnerResponse (which also uses ray.get) is that returnCode is checked before _collectRunnerResponse is called.

**Review by PaulTalbot-INL — APPROVED** [review]

**Pull request merged**